File: Project/Distributed intrusion detection/artifacts/api-server/src/routes/scan.ts
import { Router, type IRouter } from "express";
import { db, networkLogsTable, alertsTable } from "@workspace/db";
import { ScanTrafficBody, ScanTrafficResponse } from "@workspace/api-zod";
import { requireApiKey } from "../middlewares/apiKey";
import { buildAlertDescription } from "../lib/detectionService";
import { detectTraffic, type TrafficInput } from "../lib/mlDetectionClient";
import { PacketScanBody, packetToTrafficInput } from "../lib/packetIngestion";

const router: IRouter = Router();

async function scanAndPersist(input: TrafficInput) {
  const detection = await detectTraffic(input);

  const [log] = await db
    .insert(networkLogsTable)
    .values({
      sourceIp: input.sourceIp,
      destinationIp: input.destinationIp,
      protocol: input.protocol,
      data: input.data,
      status: detection.prediction,
      nodeId: input.nodeId,
      bytesSent: input.bytesSent ?? 0,
      confidenceScore: detection.confidenceScore,
    })
    .returning();

  let alertId: number | null = null;
  if (detection.prediction === "malicious" && detection.threatType) {
    const [alert] = await db
      .insert(alertsTable)
      .values({
        title: detection.threatType,
        description: buildAlertDescription(detection.threatType, input.sourceIp, detection.confidenceScore),
        severity: detection.severity,
        sourceIp: input.sourceIp,
        nodeId: input.nodeId,
        resolved: false,
        logId: log!.id,
      })
      .returning();
    alertId = alert?.id ?? null;
  }

  return ScanTrafficResponse.parse({
    logId: log!.id,
    prediction: detection.prediction,
    confidenceScore: detection.confidenceScore,
    threatType: detection.threatType,
    alertId,
    message:
      detection.prediction === "malicious"
        ? `⚠ Threat detected: ${detection.threatType} (${Math.round(detection.confidenceScore * 100)}% confidence)`
        : `✓ Traffic classified as benign (${Math.round(detection.confidenceScore * 100)}% confidence)`,
  });
}

/**
 * POST /api/scan
 *
 * Accepts raw or structured network traffic, runs the detection engine,
 * persists the log (with confidence score), auto-creates an alert if
 * malicious, and returns the full result.
 *
 * Protected by API key when DIDS_API_KEY env var is set.
 */
router.post("/scan", requireApiKey, async (req, res): Promise<void> => {
  const parsed = ScanTrafficBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const input = parsed.data;

  res.json(await scanAndPersist({
    sourceIp: input.sourceIp,
    destinationIp: input.destinationIp,
    protocol: input.protocol,
    data: input.data,
    bytesSent: input.bytesSent ?? 0,
    nodeId: input.nodeId,
  }));
});

/**
 * POST /api/scan/packet
 *
 * Accepts packet metadata or tcpdump-style text, normalizes it into the
 * standard scan payload, then uses the same ML/fallback detection workflow.
 */
router.post("/scan/packet", requireApiKey, async (req, res): Promise<void> => {
  const parsed = PacketScanBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  try {
    res.json(await scanAndPersist(packetToTrafficInput(parsed.data)));
  } catch (error) {
    res.status(400).json({ error: error instanceof Error ? error.message : "Invalid packet payload" });
  }
});

export default router;

File: Project/Distributed intrusion detection/artifacts/ml-api/app/routers/scan.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_db, get_current_user_or_internal_service
from app.models.user import User
from app.models.network_log import NetworkLog
from app.models.alert import Alert
from app.schemas.scan import ScanRequest, ScanResponse
from app.services.detection_service import DetectionService

router = APIRouter()


@router.post("/scan", response_model=ScanResponse)
async def scan(
    body: ScanRequest,
    db: AsyncSession = Depends(get_db),
    _: User | None = Depends(get_current_user_or_internal_service),
):
    """
    Run ML-based intrusion detection on incoming network traffic.
    Automatically creates an alert if malicious traffic is detected.
    """
    result = DetectionService.predict(
        source_ip=body.sourceIp,
        destination_ip=body.destinationIp,
        protocol=body.protocol,
        data=body.data,
        bytes_sent=body.bytesSent,
    )

    # Persist log
    log = NetworkLog(
        source_ip=body.sourceIp,
        destination_ip=body.destinationIp,
        protocol=body.protocol,
        data=body.data,
        node_id=body.nodeId,
        bytes_sent=body.bytesSent,
        status=result.prediction,
        confidence_score=result.confidence_score,
        threat_type=result.threat_type,
    )
    db.add(log)
    await db.flush()  # Get the log ID before creating alert

    # Auto-create alert if malicious
    alert_id: int | None = None
    if result.prediction == "malicious" and result.alert_title:
        alert = Alert(
            log_id=log.id,
            title=result.alert_title,
            description=result.alert_description or "",
            severity=result.severity,
            source_ip=body.sourceIp,
            node_id=body.nodeId,
            threat_type=result.threat_type,
        )
        db.add(alert)
        await db.flush()
        alert_id = alert.id

    await db.commit()

    message = (
        f"Threat detected: {result.threat_type}" if result.prediction == "malicious"
        else "Traffic classified as benign"
    )

    return ScanResponse(
        logId=log.id,
        prediction=result.prediction,
        confidenceScore=result.confidence_score,
        threatType=result.threat_type,
        alertId=alert_id,
        message=message,
    )
